File: frontend/src/api/launchReadiness.ts
/**
 * Stratum AI - Launch Readiness API
 *
 * Owner-only. Drives the sequential go-live wizard: phase N+1 is
 * locked until phase N is 100% complete. Each check/uncheck appends to
 * an immutable audit trail.
 */

import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiClient, ApiResponse } from './client'

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------
export interface LaunchReadinessItem {
  key: string
  title: string
  description?: string | null
  is_checked: boolean
  checked_by_user_id?: number | null
  checked_by_user_name?: string | null
  checked_at?: string | null
  note?: string | null
}

export interface LaunchReadinessPhase {
  number: number
  slug: string
  title: string
  description?: string | null
  items: LaunchReadinessItem[]
  completed_count: number
  total_count: number
  is_complete: boolean
  is_active: boolean
  is_locked: boolean
}

export interface LaunchReadinessState {
  phases: LaunchReadinessPhase[]
  current_phase_number: number
  overall_completed: number
  overall_total: number
  is_launched: boolean
}

export type LaunchReadinessEventAction =
  | 'checked'
  | 'unchecked'
  | 'phase_completed'
  | 'phase_reopened'

export interface LaunchReadinessEvent {
  id: number
  phase_number: number
  item_key?: string | null
  action: LaunchReadinessEventAction | string
  user_id?: number | null
  user_name?: string | null
  note?: string | null
  created_at: string
}

export interface ToggleItemRequest {
  itemKey: string
  checked: boolean
  note?: string
}

// -----------------------------------------------------------------------------
// API functions
// -----------------------------------------------------------------------------
export const launchReadinessApi = {
  getState: async (): Promise<LaunchReadinessState> => {
    const response = await apiClient.get<ApiResponse<LaunchReadinessState>>(
      '/console/launch-readiness'
    )
    return response.data.data
  },

  toggleItem: async ({
    itemKey,
    checked,
    note,
  }: ToggleItemRequest): Promise<LaunchReadinessState> => {
    const response = await apiClient.patch<ApiResponse<LaunchReadinessState>>(
      `/console/launch-readiness/items/${encodeURIComponent(itemKey)}`,
      { checked, note: note ?? null }
    )
    return response.data.data
  },

  getEvents: async (params: { phaseNumber?: number; limit?: number } = {}): Promise<
    LaunchReadinessEvent[]
  > => {
    const response = await apiClient.get<ApiResponse<LaunchReadinessEvent[]>>(
      '/console/launch-readiness/events',
      {
        params: {
          phase_number: params.phaseNumber,
          limit: params.limit ?? 100,
        },
      }
    )
    return response.data.data
  },
}

// -----------------------------------------------------------------------------
// React Query hooks
// -----------------------------------------------------------------------------
const stateKey = ['launch-readiness', 'state'] as const
const eventsKey = (phaseNumber?: number) =>
  ['launch-readiness', 'events', phaseNumber ?? 'all'] as const

export function useLaunchReadinessState() {
  return useQuery({
    queryKey: stateKey,
    queryFn: launchReadinessApi.getState,
    staleTime: 10 * 1000,
  })
}

export function useToggleLaunchReadinessItem() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: launchReadinessApi.toggleItem,
    onSuccess: (state) => {
      // Server already returned the fresh state; seed the cache
      // then invalidate the events log.
      queryClient.setQueryData(stateKey, state)
      queryClient.invalidateQueries({ queryKey: ['launch-readiness', 'events'] })
    },
  })
}

export function useLaunchReadinessEvents(phaseNumber?: number, limit = 100) {
  return useQuery({
    queryKey: eventsKey(phaseNumber),
    queryFn: () => launchReadinessApi.getEvents({ phaseNumber, limit }),
    staleTime: 30 * 1000,
  })
}

File: backend/app/schemas/launch_readiness.py
# =============================================================================
# ADs Growth System - Launch Readiness Schemas
# =============================================================================
"""
Pydantic schemas for the Launch Readiness wizard API.
"""

from datetime import datetime
from typing import List, Optional

from pydantic import Field

from app.base_schemas import BaseSchema


class LaunchReadinessItem(BaseSchema):
    """One checklist item inside a phase."""

    key: str
    title: str
    description: Optional[str] = None
    is_checked: bool = False
    checked_by_user_id: Optional[int] = None
    checked_by_user_name: Optional[str] = None
    checked_at: Optional[datetime] = None
    note: Optional[str] = None


class LaunchReadinessPhase(BaseSchema):
    """A phase containing its items and roll-up state."""

    number: int
    slug: str
    title: str
    description: Optional[str] = None
    items: List[LaunchReadinessItem]
    completed_count: int
    total_count: int
    is_complete: bool
    is_active: bool
    is_locked: bool


class LaunchReadinessState(BaseSchema):
    """Full launch readiness state served to the UI."""

    phases: List[LaunchReadinessPhase]
    current_phase_number: int
    overall_completed: int
    overall_total: int
    is_launched: bool


class LaunchReadinessItemUpdate(BaseSchema):
    """Payload for toggling an item's checked state."""

    checked: bool
    note: Optional[str] = Field(default=None, max_length=2000)


class LaunchReadinessEventEntry(BaseSchema):
    """Audit trail entry."""

    id: int
    phase_number: int
    item_key: Optional[str] = None
    action: str
    user_id: Optional[int] = None
    user_name: Optional[str] = None
    note: Optional[str] = None
    created_at: datetime
